### DISCUSSION TITLE
WebAssembly

### DISCUSSION THREAD

**Discussion by tostt** [cosinekitty/astronomy#348] [Ideas]
Hello,

I'm wondering if you have any thoughts regarding WebAssembly (Wasm).
I don't have experience using it, but it looks interesting.

In a nutshell, Wasm bytecode :

- can be compiled from a large number of languages (including C/C++, Kotlin, Javascript, Python, etc. More about that below).
- can be called by JavaScript code in the browser. All modern browsers support the execution of Wasm bytecode, as it's an open W3C standard.
- executes at native speed

Through a quick and cursory glance, it seems that compiling from C/C++ may give access to more mature tools. But I could be wrong.

Did you consider building Wasm bytecode (from the C version of the engine, or any other) and packaging it in your releases ?

Here is a link to the [official documentation](https://webassembly.org/getting-started/developers-guide/).

**Comment by cosinekitty** [comment]
I generally focus on providing the source code that can be used from a variety of platforms. I don't want to also maintain build environments for the platforms that could be targeted using the code. If someone wants to use Astronomy Engine in a WASM environment, they are welcome to do so. Put another way, I already have a heavy support and test matrix because of all the programming languages I support. If we start adding multiple supported platforms per programming language, the unit testing overhead becomes huge very quickly.

That said, if you end up using Astronomy Engine in a WASM environment, and there is something about the code that makes it difficult or impossible, please let me know. I would definitely want to help other people support WASM if I can make source changes that don't break any existing unit tests or demos.

**Comment by tostt** [comment]
Thank you for answering.
Adding a technical note, my understanding is that WASM bytecode is universal, i.e. platform-independant. The actual compilation is performed by the browser.
